### DISCUSSION TITLE
how to authenticate in next.js

### DISCUSSION THREAD

**Discussion by shadaryaei** [vercel/next.js#21685] [Help]
hello everyone! i'm new in next.js and Some things are vague to me.
I have a web app with the java spring backend.
as you know the authentication and persist access token and refresh token is important.
previously with CRA, I've used localStorage for persist access token and refresh token and as you know the localStorage is for client side and is undefined in server side. So what came to my mind was to use cookie. but cookie is only available in the context of getServerSideProps and getInitialProps and I don't have access to it in the component level. and in the component level, I have a series of requests that need to access token in the header of request. and i don't have access to cookie when the component to be render in the server side.

I read the API routing section in the document of next.js for persist access token and refresh token with the httpOnly flag, and I saw the profile request that the cookie has been sent over the header of the profile request. but I have a lot of request that needed the access token. Is it true that i write all of the requests in the pages/api directory and then in the component, add request to for example api/profile or something else ?

**Comment by eric-burel** [comment]
This architecture is not bad if you want more insights: https://github.com/vercel/next.js/discussions/17515

For your components, pass the cookies through props or context. Usually, you would initialize your client (be it Axios, Apollo, whatver) with this cookies you got server side in the request object.
